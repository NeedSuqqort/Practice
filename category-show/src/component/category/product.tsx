import { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import styles from "./category.module.scss";

interface productProps {
  name:string,
}

export const Product = (props: any) => {
  const [isPreviewShown, setIsPreviewShown] = useState<any>(false);
  const [isSelected, setIsSelected] = useState<any>(false);
  const [selectedItems, setSelectedItems] = props?.selecthook;
  const navigate = useNavigate();

  useEffect(() => {
    if (selectedItems[props.index]) {
      setIsSelected(true);
    }
  }, []);

  const toggle_preview = () => {
    setIsPreviewShown(!isPreviewShown);
  };

  const select_item = () => {
    console.log("item index = " + props.index);
    console.log("item name = " + props.item.name);
    setIsSelected(!isSelected);
    if (!isSelected) {
      setSelectedItems({ ...selectedItems, [props.index]: props.item.name });
    } else {
      delete selectedItems[props.index];
    }
  };

  const toggle_details = (
    index: number,
    name: any,
    imageUrl: any,
    price: any,
    description: any
  ) => {
    console.log(JSON.stringify(selectedItems));
    window.localStorage.setItem("selectedItems", JSON.stringify(selectedItems));

    console.log(`[details] click item ${index} details`);
    navigate("/item", {
      replace: false,
      state: {
        name: name,
        imageUrl: imageUrl,
        price: price,
        description: description,
      },
    });
  };

  return (
    <div
      className={`${
        isSelected ? styles.gridoxIsSelected : styles.gridboxIsNotSelected
      } ${
        isPreviewShown
          ? styles.gridboxIsPreviewed
          : styles.gridboxIsNotPreviewed
      } ${styles.gridbox}`}
      data-testid={`product-${props.index}`}
    >
      <div className={styles.productimage} id='image 1'>
        <img alt="product" src={props.item?.variantOptions[0].mainImage?.url} />
      </div>
      <div>
        <span>{props.item.name}</span>
      </div>
      <div className={styles.buttonlist}>
        <button className={styles.button} onClick={(e) => toggle_preview()}>
          Preview
        </button>
        <button
          className={styles.button}
          onClick={(e) =>
            toggle_details(
              props.index,
              props.item?.name,
              props.item?.variantOptions[0].mainImage?.url,
              props.item?.price?.formattedValue,
              props.item?.description
            )
          }
        >
          Details
        </button>
      </div>
      {isPreviewShown && (
        <div>
          <div>{props.item?.code}</div>
          <div>{props.item?.price?.formattedValue}</div>
          <div className={styles.buttonlist}>
            <button className={styles.button} onClick={(e) => select_item()}>
              {isSelected ? "Cancel" : "Select"}
            </button>
          </div>
        </div>
      )}
    </div>
  );
};
