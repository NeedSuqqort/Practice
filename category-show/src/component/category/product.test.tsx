import React, { useState } from "react";
import { render, screen, fireEvent, cleanup } from "@testing-library/react";
import { act, renderHook } from "@testing-library/react-hooks";
import { Product } from "./product";
import { waitFor } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import renderer from "react-test-renderer";


afterEach(() => {
  cleanup();
});

test("match snapshot", () => {
  const setStateMock = jest.fn();
  const useStateMock: any = (useState: any) => [useState, setStateMock];
  jest.spyOn(React, "useState").mockImplementation(useStateMock);

  const item = {
    index: 0,
    name: "iphone",
    description: "iphone is good",
    variantOptions: [
      {
        mainImage: [
          {
            url: "https://images.hktv-img.com/images/HKTV/18230/ST_Sony_Xperia1_64_JP_White_main_44326896_20200626131116_01_300.jpg",
          },
        ],
      },
    ],
  };
  const index = 0;

  const selectedItems = {};
  const setSelectedItems = jest.fn();
  const selecthook: any = [selectedItems, setSelectedItems];

  const tree = renderer
    .create(
      <Router>
        <Product item={item} index={index} selecthook={selecthook} />
      </Router>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});




test("check render product component", () => {
  const item = {
    index: 0,
    name: "iphone",
    description: "iphone is good",
    variantOptions: [
      {
        mainImage: [
          {
            url: "https://images.hktv-img.com/images/HKTV/18230/ST_Sony_Xperia1_64_JP_White_main_44326896_20200626131116_01_300.jpg",
          },
        ],
      },
    ],
  };
  const index = 0;
  const selectedItems = {};
  const setSelectedItems = jest.fn();
  const selecthook: any = [selectedItems, setSelectedItems];
  render(
    <Router>
      <Product item={item} index={index} selecthook={selecthook} />
    </Router>
  );
  const ProductElement = screen.getAllByTestId("product-0");
  waitFor(() => expect(ProductElement).toBeInTheDocument());
});

describe("test product preview button", () => {
  test("useState setState is called", () => {
    const setStateMock = jest.fn();
    const useStateMock: any = (useState: any) => [useState, setStateMock];
    jest.spyOn(React, "useState").mockImplementation(useStateMock);


    const item = {
      index: 0,
      name: "iphone",
      description: "iphone is good",
      variantOptions: [
        {
          mainImage: [
            {
              url: "https://images.hktv-img.com/images/HKTV/18230/ST_Sony_Xperia1_64_JP_White_main_44326896_20200626131116_01_300.jpg",
            },
          ],
        },
      ],
    };
    const index = 0;

    const selectedItems = {};
    const setSelectedItems = jest.fn();
    const selecthook: any = [selectedItems, setSelectedItems];

    render(
      <Router>
        <Product item={item} index={index} selecthook={selecthook} />
      </Router>
    );
    const previewButton = screen.getByText("Preview");
    fireEvent.click(previewButton);
    const SelectButton = screen.getByText("Select");
    fireEvent.click(SelectButton);
    waitFor(() => expect(setStateMock).toHaveBeenLastCalledWith(true));
  });
});

describe("test product Details button", () => {
  test("useState setState is called", () => {
    const setStateMock = jest.fn();
    const useStateMock: any = (useState: any) => [useState, setStateMock];
    jest.spyOn(React, "useState").mockImplementation(useStateMock);

    const item = {
      index: 0,
      name: "iphone",
      description: "iphone is good",
      variantOptions: [
        {
          mainImage: [
            {
              url: "https://images.hktv-img.com/images/HKTV/18230/ST_Sony_Xperia1_64_JP_White_main_44326896_20200626131116_01_300.jpg",
            },
          ],
        },
      ],
    };
    const index = 0;

    const selectedItems = {};
    const setSelectedItems = jest.fn();
    const selecthook: any = [selectedItems, setSelectedItems];

    render(
      <Router>
        <Product item={item} index={index} selecthook={selecthook} />
      </Router>
    );
    const detailsButton = screen.getByText("Details");
    fireEvent.click(detailsButton);
    waitFor(() => expect(setStateMock).toHaveBeenLastCalledWith(true));
  });
});



// describe("test product Setect button", () => {

//   test("useState setState is called", () => {


    
//     const item = {
//       index: 0,
//       name: "iphone",
//       description: "iphone is good",
//       variantOptions: [
//         {
//           mainImage: [
//             {
//               url: "https://images.hktv-img.com/images/HKTV/18230/ST_Sony_Xperia1_64_JP_White_main_44326896_20200626131116_01_300.jpg",
//             },
//           ],
//         },
//       ],
//     };
//     const index = 0;

//     const selectedItems = {};
//     const setSelectedItems = jest.fn();
//     const selecthook: any = [selectedItems, setSelectedItems];
//     let mockisPreviewShown = true
//     render(
//       <Router>
//         <Product item={item} index={index} selecthook={selecthook} />
//       </Router>
//     );
    
//     const SelectButton = screen.getByText("Select");
//     waitFor(() => expect(SelectButton).toBeInTheDocument());
//   });
// });