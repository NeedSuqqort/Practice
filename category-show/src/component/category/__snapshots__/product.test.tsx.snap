// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match snapshot 1`] = `
<div
  className="gridboxIsNotSelected gridboxIsNotPreviewed gridbox"
  data-testid="product-0"
>
  <div
    className="productimage"
  >
    <img
      alt="product"
    />
  </div>
  <div>
    <span>
      iphone
    </span>
  </div>
  <div
    className="buttonlist"
  >
    <button
      className="button"
      onClick={[Function]}
    >
      Preview
    </button>
    <button
      className="button"
      onClick={[Function]}
    >
      Details
    </button>
  </div>
</div>
`;
